Round relative time interval arguments

bogdan-szabo-sociomantic opened the pull request.

christoph-wasicki-sociomantic commented on relnotes/timedivision.feature.md:
```diff
@@ -0,0 +1,10 @@
+
+`ocean.text.arguments.TimeIntervalArgs`
+
+Relative values (eg. 1m, 4h) and 'now' can be rounded by [m]inutes, [h]ours and [d]ays.
```
"rounded" here means actually floored, right? E.g. `-t 1h/h 1h` at 15:59` would also start at `14:00`.

bogdan-szabo-sociomantic replied:
`-t 1h 1h` at `15:59` means `14:59 to 15:59` 

`-t 1h/h 1h` at `15:59` means `14:00 to 15:00` 

so, yes

christoph-wasicki-sociomantic commented on relnotes/timedivision.feature.md:
```diff
@@ -0,0 +1,10 @@
+
+Relative values (eg. 1m, 4h) and 'now' can be rounded by [m]inutes, [h]ours and [d]ays.
+
+      eg. -t 1h/h 1h = if the command is executed at 15:23:23 then
```
Minor: I did not know the format of time intervals and was confused before @bogdan-szabo-sociomantic explained it to me. The doc could repeat the meaning here `<time_start>/<floor_to_time_unit> [time_end]` or use an example that's less ambiguous (e.g. `4h/h 2h`).

bogdan-szabo-sociomantic replied:
done

liza-churkyn-sociomantic commented on src/ocean/text/arguments/TimeIntervalArgs.d:
```diff
@@ -127,6 +143,50 @@ public istring validateTimeIntervalArgs ( Arguments args )
+
+    Params:
+        op = the math operation performed tot the reference time.
+             Only "+", "-" is permitted.
```
`tot` => `to`

liza-churkyn-sociomantic commented on src/ocean/text/arguments/TimeIntervalArgs.d:
```diff
@@ -388,6 +434,71 @@ unittest
+    Test.test!("==")(interval.begin, dummy_time);
+
+    timeToUnixTime("2019-04-01 15:26:31", dummy_time, dummy_conv);
+    Test.test!("==")(interval.end, dummy_time);
```
you have both tests for hour and minutes, one test with a `/d` (day) would also be good

bogdan-szabo-sociomantic replied:
done

bogdan-szabo-sociomantic commented:
ping @matthias-wende-sociomantic

matthias-wende-sociomantic commented on relnotes/timedivision.feature.md:
```diff
@@ -0,0 +1,13 @@
+Relative values (eg. 1m, 4h) and 'now' can be rounded by [m]inutes, [h]ours and [d]ays.
+The new relative format is [time_start][time unit]/[floor_to_time_unit] or now/[floor_to_time_unit].
+
+      eg. -t 1h/h 1h = if the command is executed at 15:23:23 then
```
The example is different to what the description says.

bogdan-szabo-sociomantic replied:
how is it different?

matthias-wende-sociomantic commented on src/ocean/text/arguments/TimeIntervalArgs.d:
```diff
@@ -86,12 +100,14 @@ public void setupTimeIntervalArgs ( Arguments args, bool required )
-            "an iso1806 date (YYYY-MM-DD)");
+            "an iso1806 date (YYYY-MM-DD)\n\t\n\t"~
+            "the relative values and 'now' can be rounded by [m]inutes, [h]ours and [d]ays.\n\t" ~
+            "eg. 1h/h will set the minutes and seconds to `0`");
```
I think the help should fully document how to use the round feature.

bogdan-szabo-sociomantic replied:
I'm not sure how many details should we add here... If there is too much information, people will might just not read it

tiyash-basu-sociomantic replied:
From https://github.com/sociomantic-tsunami/ocean/pull/712/files#r287270639, I'd recommend changing `rounded` to `floored` here and everywhere else in this context.

I agree that the flooring feature needs to be documented. The syntax is already there, perhaps how to use the syntax should be added.

The more I look at it, the more I think that the rounding unit should be a separate command line argument of its own. Squashing two different inputs into one argument should be avoided as much as possible.

bogdan-szabo-sociomantic replied:
we discussed with @scott-gibson-sociomantic and @liza-churkyn-sociomantic and this feature was designed to work like grafana intervals

matthias-wende-sociomantic reviewed (commented):
First review. Two minor documentation comments and could you please provide a description. I'll review the code soon.

matthias-wende-sociomantic commented on src/ocean/text/arguments/TimeIntervalArgs.d:
```diff
@@ -457,54 +581,263 @@ unittest
+public bool isRelativeTime ( cstring value )
 {
-    if (value == "")
+    if ( value == "" )
```
This is a stylistic change and is not related to the commit.

matthias-wende-sociomantic commented on src/ocean/text/arguments/TimeIntervalArgs.d:
```diff
@@ -457,54 +581,263 @@ unittest
+
+******************************************************************************/
+
+char getTimeUnit ( cstring value )
```
missing protection attribute

matthias-wende-sociomantic commented on src/ocean/text/arguments/TimeIntervalArgs.d:
```diff
@@ -457,54 +581,263 @@ unittest
 /******************************************************************************
 
-    Checks if a string can be converted to a time interval
+    Checks if a string can be converted to a time stamp
```
That's not a very accurate definition of the function. What about `Checks if a string is a valid relative time` and maybe add an example.

bogdan-szabo-sociomantic replied:
I think there are enough examples in the unittests

matthias-wende-sociomantic commented on src/ocean/text/arguments/TimeIntervalArgs.d:
```diff
@@ -457,54 +581,263 @@ unittest
+    Test.test!("==")(isRelativeTime(""), false);
+    Test.test!("==")(isRelativeTime("1s"), true);
+    Test.test!("==")(isRelativeTime("1m"), true);
+    Test.test!("==")(isRelativeTime("2m"), true);
```
This test seems to indirectly test the `toLong` function but doesn't add any valuable new test to `isRelativeTime`.

matthias-wende-sociomantic replied:
actually that's true for most of the tests here that either test `toLong` or `isValidTimeUnit` indirectly.

matthias-wende-sociomantic commented on src/ocean/text/arguments/TimeIntervalArgs.d:
```diff
@@ -457,54 +581,263 @@ unittest
+
+******************************************************************************/
+
+public bool isDividedTime ( cstring value )
```
why is this public?

matthias-wende-sociomantic commented on src/ocean/text/arguments/TimeIntervalArgs.d:
```diff
@@ -457,54 +581,263 @@ unittest
 ******************************************************************************/
 
-public bool isTimeInterval ( cstring value )
+public bool isRelativeTime ( cstring value )
```
Why is this a public function?

matthias-wende-sociomantic commented on src/ocean/text/arguments/TimeIntervalArgs.d:
```diff
@@ -457,54 +581,263 @@ unittest
+
+******************************************************************************/
+
+public bool isValidTimeUnit ( char unit )
```
public?

matthias-wende-sociomantic commented on src/ocean/text/arguments/TimeIntervalArgs.d:
```diff
@@ -457,54 +581,263 @@ unittest
+
+******************************************************************************/
+
+public cstring removeTimeDivision ( cstring value )
```
Again why is this a public method?

matthias-wende-sociomantic commented on src/ocean/text/arguments/TimeIntervalArgs.d:
```diff
@@ -457,54 +581,263 @@ unittest
+
+******************************************************************************/
+
+long roundTime ( long value, char unit )
```
Missing protection attribute.

matthias-wende-sociomantic reviewed (commented):
This PR seems to do more then what it says in the title. You might want to explain why you refactored the module and may split the changes into better structured smaller commits.

Also it seems that exposing all methods as public is not necessary.

bogdan-szabo-sociomantic commented:
@ben-palmer-sociomantic can you have a look at this pr?

bogdan-szabo-sociomantic committed:
Rename isTimeInterval to isRelativeTime

ben-palmer-sociomantic commented on src/ocean/text/arguments/TimeIntervalArgs.d:
```diff
@@ -127,6 +127,46 @@ public istring validateTimeIntervalArgs ( Arguments args )
+
+    if ( isTimeInterval(value) )
+    {
+        static if ( op == "+" )
```
Maybe add a static assert on the value of `op`?

ben-palmer-sociomantic commented on src/ocean/text/arguments/TimeIntervalArgs.d:
```diff
@@ -142,48 +182,33 @@ public TimestampInterval processTimeIntervalArgs ( Arguments args )
     enforce(num_args >= 1 && num_args <= 2, "Not enough arguments provided");
     bool include_end_date = !args["time-interval-exclude"].set;
 
-    long begin, end;
```
Should the docs for this method be updated as the functionality has changed?

bogdan-szabo-sociomantic replied:
I think it's better to update the release notes

ben-palmer-sociomantic commented on src/ocean/text/arguments/TimeIntervalArgs.d:
```diff
@@ -517,7 +517,7 @@ unittest
 ******************************************************************************/
 
-public bool isTimeInterval ( cstring value )
+private bool isRelativeTime ( cstring value )
```
This is a change to the public API that would need to be in a major release.

bogdan-szabo-sociomantic replied:
I don't think that there is anyone who uses this function...

ben-palmer-sociomantic replied:
This is an open source library so we can't actually say that for sure.

gavin-norman-sociomantic replied:
Correct. A change like this must always be considered a breaking change. If the function must be renamed, consider using a deprecation.

bogdan-szabo-sociomantic replied:
done

ben-palmer-sociomantic commented on src/ocean/text/arguments/TimeIntervalArgs.d:
```diff
@@ -86,12 +100,14 @@ public void setupTimeIntervalArgs ( Arguments args, bool required )
         .aliased('t')
         .params(1, 2)
-        .help("Specify a time interval. It 1 or 2 values, where they can be:\n\t" ~
+        .help("Specify a time interval. It takes 1 or 2 values, where they can be:\n\t" ~
```
This documentation change is not related to this commit.

ben-palmer-sociomantic reviewed (commented):
I had a couple of comments.

liza-churkyn-sociomantic commented on src/ocean/text/arguments/TimeIntervalArgs.d:
```diff
@@ -356,7 +382,7 @@ unittest
 
     Test.test!("==")(interval.begin, TEST_TIME_NOW - 60);
-    Test.test!("==")(interval.end, TEST_TIME_NOW + 59);
+    Test.test!("==")(interval.end, TEST_TIME_NOW - 1);
```
This commit contains two changes: 1. moving the duplicate part out 2. functionality change, where in case of 2 arguments `end` becomes relating to `begin` and not relative to `now`. Due to the functionality change you had to update the unit test.  

So, you could split this commit to two: the 1st commit would contain the unit test update `Test.test!("==")(interval.end, TEST_TIME_NOW - 1);` and the functionality change:
```
    begin = time(null) - parseTimeInterval(begin_arg);	
    end = begin + parseTimeInterval(end_arg);
```

Instead of the previous:
```
    begin = time(null) - parseTimeInterval(begin_arg);	
    end = time(null) + parseTimeInterval(end_arg);
```

The 2nd commit can contain the rest.

liza-churkyn-sociomantic commented on src/ocean/text/arguments/TimeIntervalArgs.d:
```diff
@@ -200,7 +220,7 @@ public TimestampInterval processTimeIntervalArgs ( Arguments args )
     return createTimestampInterval(
         args["time-interval"].assigned[0],
-        args["time-interval"].assigned[0],
+        removeTimeDivision(args["time-interval"].assigned[0]),
```
Why are you calling `removeTimeDivision` here? Here end = begin = "non relative time"...

bogdan-szabo-sociomantic replied:
it's needed because of this case: `--time-interval now/h` which should be the same as `--time-interval now/h now`

liza-churkyn-sociomantic commented on src/ocean/text/arguments/TimeIntervalArgs.d:
```diff
@@ -226,7 +246,7 @@ private TimestampInterval createTimestampInterval ( cstring str_begin,
     if ( !isRelativeTime(str_end) )
     {
-        reference_time = parseDateString(str_end, include_end_date);
+        reference_time = parseDateString(removeTimeDivision(str_end), include_end_date);
```
Same question as above, don't we remove division only from "relative time" values?

bogdan-szabo-sociomantic replied:
it's needed because `parseDateString` does not know how to parse strings with time divisions. 

`"--time-interval now/h now/m"` is a valid case

liza-churkyn-sociomantic commented on src/ocean/text/arguments/TimeIntervalArgs.d:
```diff
@@ -553,7 +645,222 @@ unittest
+******************************************************************************/
+
+private bool isValidTimeUnit ( char unit )
+{
```
Taking out the `unit == 's' || unit == 'm' || unit == 'h' || unit == 'd'` code and putting it a separate function `isValidTimeUnit` can be in its own commit (before this commit)

bogdan-szabo-sociomantic replied:
Should I go in that many details with the commits?

bogdan-szabo-sociomantic committed:
Fix help message for time arguments

bogdan-szabo-sociomantic committed:
Make end argument to be relative to the begin argument instead of the current timestamp

bogdan-szabo-sociomantic committed:
Move the interval initialization to separate function

bogdan-szabo-sociomantic committed:
Add isValidTimeUnit to check if a char is a valid time unit

bogdan-szabo-sociomantic committed:
Move duplicated parsing code to a function

bogdan-szabo-sociomantic committed:
Implement yesterday keyword as relative time

bogdan-szabo-sociomantic committed:
Round relative time interval arguments

gavin-norman-sociomantic commented:
(Accidentally clicked on CLOSE, sorry!)

gavin-norman-sociomantic commented:
@bogdan-szabo-sociomantic this contains a breaking change now, right? If so, it'll need a `x.breaking.md` release note please.

ben-palmer-sociomantic commented on src/ocean/text/arguments/TimeIntervalArgs.d:
```diff
@@ -127,9 +143,72 @@ public istring validateTimeIntervalArgs ( Arguments args )
+        }
+        else
+        {
+            static assert(true, "Only `+` and `-` operations are supported.");
```
Should this be `false`?

bogdan-szabo-sociomantic replied:
yes, thanks

ben-palmer-sociomantic commented on src/ocean/text/arguments/TimeIntervalArgs.d:
```diff
@@ -403,7 +567,7 @@ unittest
 ******************************************************************************/
 
-public long parseTimeInterval ( cstring value )
+private long parseTimeInterval ( cstring value )
```
This is another breaking change in the API that needs to be addressed.

bogdan-szabo-sociomantic replied:
reverted to public

ben-palmer-sociomantic commented on src/ocean/text/arguments/TimeIntervalArgs.d:
```diff
@@ -457,54 +621,266 @@ unittest
+        return value;
+    }
+
+    return value[0..$-2];
```
Should we do a check on array length here?

bogdan-szabo-sociomantic replied:
not really, but I added one.

ben-palmer-sociomantic reviewed (commented):
Couple more comments from me.

gavin-norman-sociomantic commented on relnotes/timedivision.breaking.md:
```diff
@@ -0,0 +1,5 @@
+
+`ocean.text.arguments.TimeIntervalArgs`
+
+isTimeInterval ( cstring value ) is now deprecated
```
A method being deprecated itsn't a breaking change.

ben-palmer-sociomantic replied:
Agreed. Should be `timedivision.deprecation.md`.

bogdan-szabo-sociomantic replied:
done :)

gavin-norman-sociomantic commented:
@bogdan-szabo-sociomantic when you're back, let's look at this PR. I can't believe it's impossible to implement this without a breaking change. Breaking changes should be an absolute last resort. Let's find a way to merge this into v4.x.x.

ben-palmer-sociomantic reviewed (commented):
Tests seem to be failing now.

bogdan-szabo-sociomantic commented:
@ben-palmer-sociomantic This started to happen after I rebased to v5... it looks like the prometheus module has a compiling error: 

```
./src/ocean/util/prometheus/collector/StatFormatter.d(216): Error: template instance ocean.util.prometheus.collector.StatFormatter.getSanitized!float error instantiating
```

tiyash-basu-sociomantic commented:
@bogdan-szabo-sociomantic The prometheus module is D1, and it was added by neptune to the D2-only v5.x.x branch without applying d1to2fix on it. This is causing the error that you saw. I'd recommend waiting until its fixed (https://github.com/sociomantic-tsunami/ocean/pull/715).

gavin-norman-sociomantic commented:
#715 is fixed now. But this PR should be based on v4x.x. As I said before, there's no reason it needs to be a breaking change.

gavin-norman-sociomantic commented:
#715 is fixed now. But this PR should be based on v4x.x. As I said before, there's no reason it needs to be a breaking change.

bogdan-szabo-sociomantic commented:
@gavin-norman-sociomantic I rebased to 4.x.x 

I am not sure if there are any breaking changes left. All the new functions are private, and for `isTimeInterval` that I renamed to `isRelativeTime`, there is an alias with the old name and marked as deprecated

ben-palmer-sociomantic commented:
This doesn't seem to be running the tests for some reason.

gavin-norman-sociomantic commented on src/ocean/text/arguments/TimeIntervalArgs.d:
```diff
@@ -561,18 +561,21 @@ public bool isTimeInterval ( cstring value )
 }
 
+/// ditto
+public deprecated alias isRelativeTime isTimeInterval;
```
Deprecation message?

gavin-norman-sociomantic commented on relnotes/timedivision.deprecation.md:
```diff
@@ -0,0 +1,5 @@
+### Round relative time interval arguments
```
The commit message for "Rename isTimeInterval to isRelativeTime" should mention the reason for this change. (Currently, while the code change looks fine, there's no context as to why we might want to merge this change.)

gavin-norman-sociomantic commented on src/ocean/text/arguments/TimeIntervalArgs.d:
```diff
@@ -225,15 +225,7 @@ public TimestampInterval processTimeIntervalArgs ( Arguments args )
-    {
-        end = parseTimeArgument!("+")(end_arg, begin, include_end_date);
-    }
+    end = parseTimeArgument!("+")(end_arg, begin, include_end_date);
```
The commit message "Make end argument to be relative to the begin argument instead of the current timestamp" should describe the rationale for the change.

gavin-norman-sociomantic commented on src/ocean/text/arguments/TimeIntervalArgs.d:
```diff
@@ -198,9 +198,10 @@ public long parseTimeArgument ( cstring op ) ( cstring value, long reference_tim
     an iso8601 date, a timestamp, `now` or `yesterday`.
 
     If only one argument is provided, the second one will be infered:
+        - `yesterday` it's rewriten as `1d/d 1d`
```
The commit "Implement yesterday keyword as relative time" appears to be changing some previous behaviour, rather than implementing a new behaviour. The commit message should describe what the old behaviour was and the reason for changing it.

gavin-norman-sociomantic replied:
It sounds like this could potentially be a breaking change, but without under understanding the old behaviour and the rationale behind changing it, it's difficult to say.

If it does turn out to be a breaking change, we could introduce a new argument, in addition to `"yesterday"`.

gavin-norman-sociomantic reviewed (commented):
Partial review. General feedback: this PR is difficult to review, for the following reasons:
1. It mixes a lot of seemingly unrelated changes. Or are they related? This conjures a feeling of unease for the reviewer.
2. There's virtually no documentation of the reason for the changes. Without understanding why a change is proposed, it's very difficult to make a judgment on it.

Suggestion:
1. Split the PR into several smaller ones, each consisting of purely strictly related changes.
2. As mentioned in my review comments, enhance the commit messages to explain what the old behaviour was, what's changing, and why the changes are being made.
